### DISCUSSION TITLE
blobstore: why doesn't the scheduler periodically check whether the task's leaseExpiredS has expired?

### DISCUSSION THREAD

**Discussion by jianwei1216** [cubefs/cubefs#3465] [Q&A] [answered]
balancer、disk_droper、manual_migrater、 disk_repair  Tasks

After these tasks are assigned by BlobNode， 
BlobNode will periodically renew the lease of these Tasks from Scheduler to avoid expiration.
This allows blobnode to exclusively use these tasks to avoid data migration or repair conflicts.

When the blobnode and scheduler network is abnormal （blobnode ==XXX==> scheduler leader)，
blobnode will immediately stop all running tasks.

```go
func (tm *TaskRunnerMgr) renewalTask() {
	aliveTasks := tm.GetAliveTasks()
	if len(aliveTasks) == 0 {
		return
	}

	span, ctx := trace.StartSpanFromContext(context.Background(), "renewalTask")
	ret, err := tm.renewalCli.RenewalTask(ctx, &scheduler.TaskRenewalArgs{IDC: tm.idc, IDs: aliveTasks})
	if err != nil {
		span.Errorf("renewal task failed and stop all runner: err[%+v]", err)
		tm.StopAllAliveRunner()  // here, stop all running task !!!
		return
	}
```

but the scheduler does not reclaim these Tasks

Question: 

blobnode == periodically acquire task ==> scheduler
blobnode ==> periodically renew task ==> scheduler ==> update task.leaseExpiredTime
blobnode ==> renew task failed when network abnoraml ==> stop all running tasks
BUT scheduler ==> not periodically check task.leaseExpiredTime

How can we re-execute tasks that appear to have been allocated by the scheduler but have actually been released by the blobnode?

@Cloudstriff

**Comment by Cloudstriff** [comment, marked as answer]
When a Blobnode renews its lease, the corresponding task's expiration time in the queue will be extended, and at this time, the queue pop operation will not pop that task. However, when the Blobnode fails to renew the lease (such as self-release, node restart, etc.), the task in the queue will expire, and when other Blobnode  acquire tasks, the expired tasks will be prioritized for popping and returning.

**Reply by jianwei1216** [reply]
```
// AddPreparedTask add prepared task
func (q *WorkerTaskQueue) AddPreparedTask(idc, taskID string, wtask WorkerTask) {
	q.mu.Lock()
	defer q.mu.Unlock()

	idcQueue, ok := q.idcQueues[idc]
	if !ok {
		idcQueue = NewQueue(q.leaseExpiredS)
		q.idcQueues[idc] = idcQueue
	}
	err := idcQueue.Push(taskID, wtask)
	if err != nil {
		panic("unexpect add prepared task fail:" + err.Error())
	}
}
```


```
// Push push message to queue id is uniquely identifies。
func (q *Queue) Push(id string, msg interface{}) error {
	q.mu.Lock()
	defer q.mu.Unlock()

	if _, ok := q.msgs[id]; ok {
		return errExistingMessageID
	}

	m := &msgEx{
		id:    id,
		state: msgStateTodo,
		msg:   msg,
	}
	elem := q.todo.PushBack(m)
	q.msgs[id] = elem

	return nil
}
```



```
// Pop  fetch a msg from queue。
func (q *Queue) Pop() (string, interface{}, bool) {
	q.mu.Lock()
	defer q.mu.Unlock()

	now := time.Now()
	for ele := q.doing.Front(); ele != nil; ele = ele.Next() {
		m := ele.Value.(*msgEx)
		if m.deadline.Before(now) {
			m.deadline = now.Add(q.msgTimeout)
			return m.id, m.msg, true
		}
	}

	// no timeout msg in doing ,fetch from todo
	if q.todo.Len() == 0 {
		return "", nil, false
	}
	elem := q.todo.Front()
	q.todo.Remove(elem)

	m := elem.Value.(*msgEx)
	m.state = msgStateDoing
	m.deadline = now.Add(q.msgTimeout)

	elem = q.doing.PushFront(m)
	q.msgs[m.id] = elem

	return m.id, m.msg, true
}
```

**Reply by jianwei1216** [reply]
Good idea, it is hard to find if you don't read the code carefully
